### PR TITLE
Enhance resilience of automatic actions

### PR DESCRIPTION
**Pull request by cconard96**
## Checklist before requesting a review

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] This change requires a documentation update.

## Description

I assume there won't be time for a proper rework of automatic actions for GLPI 12, so I am attempting some smaller enhancements to hopefully reduce the long-standing issues of tasks getting stuck "running".

For a proper rework, it seems like a good use of [Symfony Scheduler](https://outline.teclib.com/doc/symfony-scheduler-for-automatic-actions-cepfLh52p3).

Changes:
- Replace `declare(ticks=1)` with recommended replacement `pcntl_async_signals(true);` if `pcntl` available.
- Add shutdown handler for normal/graceful terminations including when exceptions are thrown or the max execution time is exceeded. `pctnl` still can be used to handle external terminations like SIGTERM if the extension is available.
- Replace basic `lastrun + frequency` calculation for determining tasks that are ready to run with a `next_run` field that allows for more flexibility.
- Add an error counter to track consecutive failures and implemented an exponential backoff strategy.
- If a task reaches a maximum number of consecutive errors, the task will be put in a new error state which would require administrator intervention to reset.

### CODE REVIEW

**Comment by cconard96** [comment]
TODO:
MariaDB Galera and probably other clusters do not support use of `GET_LOCK`?
https://forum.glpi-project.org/viewtopic.php?id=287153

Maybe there is a quick/easy way to implement an alternative locking mechanism. Symfony Lock may be a simple enough implementation for GLPI 12.

**Comment by cconard96** [comment]
I added 3 temporary automatic actions to test a few failure conditions.

```sql
INSERT INTO glpi_crontasks (itemtype, name, frequency, param, state, mode, allowmode, hourmin, hourmax, logs_lifetime, lastrun, lastcode, comment, date_mod, date_creation, error_count, next_run) VALUES ('CronTask', 'exception', 60, null, 3, 1, 3, 0, 24, 30, '2026-04-15 18:33:00', null, null, null, null, 0, null);
INSERT INTO glpi_crontasks (itemtype, name, frequency, param, state, mode, allowmode, hourmin, hourmax, logs_lifetime, lastrun, lastcode, comment, date_mod, date_creation, error_count, next_run) VALUES ('CronTask', 'fatalError', 60, null, 3, 1, 3, 0, 24, 30, '2026-04-15 18:41:00', null, null, null, null, 0, null);
INSERT INTO glpi_crontasks (itemtype, name, frequency, param, state, mode, allowmode, hourmin, hourmax, logs_lifetime, lastrun, lastcode, comment, date_mod, date_creation, error_count, next_run) VALUES ('CronTask', 'maxExecutionTime', 60, null, 1, 1, 3, 0, 24, 30, '2026-04-15 18:44:00', null, null, null, null, 0, null);

```

In all cases, the task state was reset back to waiting and incremented the error count. After 5 consecutive errors, the task state was changed to ERROR.

**Comment by cconard96** [comment]
> TODO: MariaDB Galera and probably other clusters do not support use of `GET_LOCK`? https://forum.glpi-project.org/viewtopic.php?id=287153
> 
> Maybe there is a quick/easy way to implement an alternative locking mechanism. Symfony Lock may be a simple enough implementation for GLPI 12.

Not strictly related to automatic actions. Should be done in a different PR.

File: src/CronTask.php [hunk 1]
```diff
@@ -1985,4 +2053,58 @@ public static function showSearchStatusArea(): void
...
+    }
+
+    /**
+     * A cron task that always throws an exception, used to test error handling of cron tasks
+     *
+     * @param self $task for log
+     *
+     * @return int
+     * @used-by self
+     * @todo REMOVE BEFORE MERGING PULL REQUEST, this is only for testing purpose
```
Comment (AdrienClairembault) [inline src/CronTask.php]: This was not removed.

File: src/CronTask.php [hunk 2]
```diff
@@ -1985,4 +2053,58 @@ public static function showSearchStatusArea(): void
...
+    }
+
+    /**
+     * A cron task that always throws a Symfony fatal error, used to test error handling of cron tasks
+     *
+     * @param self $task for log
+     *
+     * @return int
+     * @used-by self
+     * @todo REMOVE BEFORE MERGING PULL REQUEST, this is only for testing purpose
```
Comment (AdrienClairembault) [inline src/CronTask.php]: This was not removed.

File: src/CronTask.php [hunk 3]
```diff
@@ -1985,4 +2053,58 @@ public static function showSearchStatusArea(): void
...
+            'file' => __FILE__,
+            'line' => __LINE__,
+        ]);
+    }
+
+    /**
+     * A cron task that always exceeds the max execution time, used to test handling of long-running cron tasks
+     * @param CronTask $task
+     * @return int
+     * @todo REMOVE BEFORE MERGING PULL REQUEST, this is only for testing purpose
```
Comment (AdrienClairembault) [inline src/CronTask.php]: This was not removed.

File: src/CronTask.php [hunk 4]
```diff
@@ -498,7 +566,7 @@ public function getNeedToRun(int $mode = 0, string $name = ''): bool
             $WHERE[] = [
                 'OR' => [
                     'lastrun'   => null,
-                    new QueryExpression(QueryFunction::unixTimestamp('lastrun') . ' + ' . $DB::quoteName('frequency') . ' <= ' . QueryFunction::unixTimestamp()),
+                    new QueryExpression(QueryFunction::unixTimestamp('next_run') . ' <= ' . QueryFunction::unixTimestamp()),
```
Comment (AdrienClairembault) [inline src/CronTask.php]: Isn't there a risk for existing tasks with a non null last-run to never be executed since next_run will be null after the migration?

File: src/CronTask.php [hunk 5]
```diff
@@ -249,8 +267,11 @@ public function start(): bool
         }
 
         if (isCommandLine() && function_exists('pcntl_signal')) {
+            pcntl_async_signals(true);
             pcntl_signal(SIGTERM, [$this, 'signal']);
         }
+        // Shutdown handler will be used for normal terminations (exit, exception, etc)
+        register_shutdown_function($this->handleAbnormalTermination(...));
```
Comment (AdrienClairembault) [inline src/CronTask.php]: Comment says normal termination but method name says abnormal, this seems unwanted (on normal termination run abnormal handler?).

File: src/Glpi/System/Status/StatusChecker.php [hunk 6]
```diff
@@ -478,8 +478,18 @@ public static function getCronTaskStatus(bool $public_only = true): array
                 foreach ($stuck_crontasks as $ct) {
                     $status['stuck'][] = $ct['name'];
                 }
+                $errored_crontasks = array_filter($crontasks, static fn($crontask) => $crontask['state'] === CronTask::STATE_ERROR);
+                foreach ($errored_crontasks as $ct) {
+                    $status['errored'][] = $ct['name'];
+                }
                 $status['status'] = count($status['stuck']) ? self::STATUS_PROBLEM : self::STATUS_OK;
```
Comment (AdrienClairembault) [inline src/Glpi/System/Status/StatusChecker.php]: Shouldn't it also report errors if count($errored_crontasks) is not zero? It only looks at the stuck tasks here.*

This mean multiple tasks in error would be reported as STATUS_OK.

File: src/CronTask.php [hunk 7]
```diff
@@ -342,10 +395,25 @@ public function end(?int $retcode, int $log_state = CronTaskLog::STATE_STOP): bo
...
 
+        if (is_null($retcode) || $log_state === CronTaskLog::STATE_ERROR) {
+            $this->fields['error_count'] = ($this->fields['error_count'] ?? 0) + 1;
+        } else {
+            $this->fields['error_count'] = 0;
+        }
+
+        if ($this->fields['error_count'] >= self::MAX_ERROR_COUNT) {
+            $this->fields['state'] = self::STATE_ERROR;
+            $this->sendNotificationOnError();
```
Comment (AdrienClairembault) [inline src/CronTask.php]: `sendNotificationOnError` is already called after `end`:

Image

Is it normal to call it again? It will be called twice in case of errors.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Lint is failing.

File: src/CronTask.php [hunk 8]
```diff
@@ -215,22 +219,36 @@ public static function getUsedItemtypes(): array
...
+        // End of this task
+        $this->end(null);
 
-            // End of this cron
-            $_SESSION["glpicronuserrunning"] = '';
-            self::release_lock();
-            Toolbox::logInFile('cron', __('Action aborted') . "\n");
+        // End of this cron
+        $_SESSION["glpicronuserrunning"] = '';
+        self::release_lock();
```
Comment (AdrienClairembault) [inline src/CronTask.php:251]: Aren't MySQL lock session based and released anyway when the process crash?

**Review by AdrienClairembault — COMMENTED** [review]
As a subjective comment, I don't like the original code from GLPI here that rely on signals and terminations handlers.

I would prefer replacing it by something simpler instead of building it up with new safeties (and increased complexity).

My intuition tells me that signals are overkill here and simple dates would handle this just fine.

I asked opus for some feedback to get a second opinion and it seems to agree (take it with a grain of salt, as AIs can make mistakes of course):

Image

----

This is subjective so don't take it as a blocking comment for the merge. Just sharing my opinion :)

File: src/CronTask.php (hunk 1 above)
Comment (cconard96) [reply src/CronTask.php]: I left the temporary actions to make testing/reviewing this PR easier. I will remove them as a final step.

File: src/CronTask.php (hunk 5 above)
Comment (cconard96) [reply src/CronTask.php]: Poor choice of words.
Normal termination in the comment means from a PHP context, not in the context of the task.

File: src/CronTask.php (hunk 8 above)
Comment (cconard96) [reply src/CronTask.php:251]: Named locks are global but are *supposed* to be released when the session that created the lock ends. I didn't change locking behavior in this PR.

If there are any risks of locks being held too long, I would think it would be with the file-based locks it also uses.

**Comment by cconard96** [comment]
> This is subjective so don't take it as a blocking comment for the merge. Just sharing my opinion :)

I don't put much weight in what the bot says, but I understand where your comments are coming from.
I would prefer replacing it too, but had no hope of that happening for GLPI 12. As far as I know, there was no plan to ever change or replace this code. Since community comments constantly mention tasks not running or getting stuck, any attempt to improve the current code could be nice as a short-term solution. I would prefer handing off most of the responsibility to Symfony and letting admins choose better components/services that meet their needs (Redis for locking. Redis, PHP AMQP, or Amazon SQS for messaging queues). 

For the watcher task, it only runs once a day by default and the logic it uses isn't the best and it itself isn't reliable.
- For a task that runs every minute like sending notifications, it could legitimately need to run for more than 2 minutes depending on the mail server, the amount of mail to send, etc.
- The watcher task only alerts admins about potentially stuck tasks, not reset them, and that alert logic relies on the notification sending task which is the most frequent one I have seen mentioned as not working/stuck.

The only "reliable" way to check to stuck tasks is to use an external service with the GLPI status checker. However there are cases where tasks are constantly getting "stuck" even temporarily, so admins may simply ignore them. Having a distinct error state shows to the admins that the task was indeed seen not completely properly **and** that GLPI retried the task multiple time, so admin intervention is definitely needed.

File: src/CronTask.php [hunk 9]
```diff
@@ -249,8 +267,11 @@ public function start(): bool
         }
 
         if (isCommandLine() && function_exists('pcntl_signal')) {
+            pcntl_async_signals(true);
             pcntl_signal(SIGTERM, [$this, 'signal']);
         }
+        // Shutdown handler will be used for normal terminations (exit, exception, etc)
```
Comment (cedric-anne) [inline src/CronTask.php]:
```suggestion
        // Shutdown handler will be used to catch unexpected PHP script execution ending (unrecoverable errors, uncaught errors, unexpected exit, etc).
```

File: src/CronTask.php [hunk 10]
```diff
@@ -342,10 +395,25 @@ public function end(?int $retcode, int $log_state = CronTaskLog::STATE_STOP): bo
             return false;
         }
 
+        if (is_null($retcode) || $log_state === CronTaskLog::STATE_ERROR) {
+            $this->fields['error_count'] = ($this->fields['error_count'] ?? 0) + 1;
+        } else {
+            $this->fields['error_count'] = 0;
+        }
```
Comment (cedric-anne) [inline src/CronTask.php:407]: Maybe we should introduce an `ABORTED` status, to be able to distinguish unexpected errors and requested abortions.
-> If PHP receives a `SIGTERM` signal, it means that the termination was requested, and therefore this abortion is not due to an error during the execution. We cannot do this on our side, it is up to the administrator to find the cause of this abortion.
-> If the scripts exits due to an uncaught exception, or due to an unrecoverable error, then this is probably due to a bug or a bad design of our code that results in a memory exhaustion.

An `ABORTED` task should be probably be reexecuted on the next run, without additional delay.

File: src/CronTask.php
Comment (cedric-anne) [inline src/CronTask.php]:
```suggestion
            'next_run' => 'NULL',
```

File: src/CronTask.php [hunk 11]
```diff
@@ -1985,4 +2053,58 @@ public static function showSearchStatusArea(): void
             );
         }
     }
+
+    public function prepareInputForUpdate($input)
+    {
+        if (isset($input['state']) && (int) $this->fields['state'] !== self::STATE_WAITING && (int) $input['state'] === self::STATE_WAITING) {
+            $input['error_count'] = 0;
+        }
```
Comment (cedric-anne) [inline src/CronTask.php:2099]: Maybe `error_count` should be manually added to the input when it is appropriate. If we later change the call to `$DB->update()` to a call to `CommonDBTM::update()` in the `CronTask::end()` method, then the `error_count` field incrementationw ill not be working anymore.

File: templates/pages/setup/crontask/crontask.html.twig [hunk 12]
```diff
@@ -33,145 +33,154 @@
...
+
+                            <div class="row flex-row">
+                                {# Last run #}
+                                {% set last_run_field %}
+                                    {% if item.fields['lastrun'] is empty %}
+                                        {{ __('Never') }}
+                                    {% else %}
+                                        {{ item.fields['lastrun']|formatted_datetime }}
+                                        {% if can_execute %}
+                                            {# ID and CSRF token fields are defined in components/form/buttons.html.twig #}
```
Comment (cedric-anne) [inline templates/pages/setup/crontask/crontask.html.twig]: No more CSRF fields in GLPI 12.0.
Same remark for other similar changes in this file.

```suggestion
                                            {# ID defined in components/form/buttons.html.twig #}
```

**Review by cedric-anne — COMMENTED** [review]
Using the `register_shutdown_function` is the best solution to handle unrecoverable errors. I am not sure it handles a process killing made by the Widows task scheduler, but, in this specific case, I am not sure we can do much.

File: src/CronTask.php (hunk 4 above)
Comment (cconard96) [reply src/CronTask.php]: It seems so. a `UNIX_TIMESTAMP(NULL)` returns null and a comparison `NULL < NUMBER` returns null. I've fixed it by wrapping the `QueryFunction::unixTimestamp('next_run')` in an `IFNULL` function.

**Comment by cedric-anne** [comment]
@cconard96 Is a this ready for a new review ?

**Comment by cconard96** [comment]
> @cconard96 Is a this ready for a new review ?

I think so, yes. I removed the automatic action methods that purposefully failed to test the error handling. I'm not really sure how else this can be functionally tested in a review though as most failures seem to be edge-cases or environment-specific. This is one of the things that may only be really tested in production.

**Review by AdrienClairembault — COMMENTED** [review]
I am still somewhat unhappy with going with another signal + shutdown handler solution.

GLPI already has a `getZombieCronTasks` method to detect stuck crons, If possible I'd rather use it to detect crashed tasks and restart them properly on the next run rather than try to recover from crashes that we can't always predict.

If everyone else is happy with it you can ignore my comment and move forward, that's just my opinion and this is not a critical subject :)

**Comment by cconard96** [comment]
> GLPI already has a getZombieCronTasks method to detect stuck crons

True, but if there is an issue with some common code, you can't rely on the `watcher` CronTask to tell you of issues with CronTasks at all. Plus if there are issues with notifications, which is probably the one I see issues with the most, you never get notified of the stuck tasks it finds.

This PR makes a proper error state where we **know** it finished in an unexpected way and is therefore different from the previous "stuck" detection where tasks somehow remain in a running state for a long time but has no idea why; The task may actually still be running.

**Review by cedric-anne — APPROVED** [review]
